### PR TITLE
MM-230 Use common panel to display search results and include # of co-ops filtered

### LINKED ISSUE #230: [Dotcoop - Misc] Display # of co-ops filtered in UI
**Please track against the clockify project [Dotcoop - Misc]**

## Description
Implement the design chosen in #232 to display the number of results found matching the text search and filter options chosen.  

Designs: https://xd.adobe.com/view/aefe2e1a-612a-4ed4-b570-0bf513927e8c-fbad/grid
**Updated designs 3/4/24:** https://xd.adobe.com/view/06927e2d-af74-4e6f-8749-35a116f28e33-53c1/grid

Note the significant additional change here to have the filter results  open in a new panel instead of underneath, unifying the design pattern between the directory panel and the filter panel.
On mobile only one panel is displayed at a time, so if a user wishes to add more than one filter they will be switched to the new panel unexpectedly which would be a bad UX. Hence the addition of the 'Apply Filters' button to enable the user to add multiple filters on mobile. 
 
## Acceptance Criteria

- The count should include initiatives with no geolocation, so with no marker on the map. 
- When all filters are clear and no text search set, it should display the number of all the initiatives in the directory. 
- The display method should work for between 0 and 999,999 initiatives. 
- It should work on all screen sizes.
- It should be enabled/disabled via map configuration.
- Palette should be configurable with other widget palettes.

### PR DESCRIPTION
**Pull request by rogup**
Closes #230 

The main change here is to generalise the dark initiatives list panel used in DirectorySidebarView, to also be used to display the list of initiatives that match a search. In that process of moving code over to the abstract class, I also spotted a lot of unnecessary duplicated code in the child classes, which I removed.

This code to create the initiatives list panel has been moved to BaseSidebarView, and a count of results has been added to its header. It should appear whenever a search is performed or a directory item is clicked.

For smaller screens with a width of < 1080 pixels, the panel only displays once 'apply filters' has been clicked, since it appears in front of the sidebar. The logic and css for detecting smaller screens has been fixed up since it was a bit inconsistent.

### CODE REVIEW

**Comment by wu-lee** [comment]
I've had a look through. More specific comments to follow.

> The UI code is a mess so I think it will be difficult. It was hard to separate my changes into separate commits due the messiness of the code, where I kept discovering different code paths during development.

Yes indeed.  Welcome to my world.

As someone who doesn't often get asked to review other people's code, I'm wondering about the following options:
- hide my befuddled expression with a thumbs-up and wave it on through
- work through the detailed changes and ask lots of perplexed and borderline passive aggressive questions in comments
- request that you rebase it splitting it out into the following categories I can discern
  - cosmetic changes (indentation, spelling, bracketing, wrapping)
  - symbol name changes
  - removal of old/duplicate code
  - changing the implementation of apparently unrelated functionality
  - a new "apply filters" button
  - changes to the way search results are displayed on small screens
  - the addition of the search result total
- rebase it myself to the same end and [flex](https://youtu.be/r6tH55syq0o) on you with my [git tips and tricks](https://www.youtube.com/watch?v=ecK3EnyGD8o)
- defer replying until you give up and merge it yourself
- give up and [learn to farm](https://youtu.be/r6tH55syq0o?t=339)

It's a bit of a dilemma. I might go with the first option, but with a dash of two, after a break for a hot cup of tea; mitigate this with an attempt to defuse tension through humour; then if I'm feeling ambitious perhaps have a crack at four, but with a reserve option of five or six if all this works out badly.

File: src/map-app/localisations.ts [hunk 1]
```diff
@@ -138,11 +145,13 @@ export const phraseBooks: PhraseBooks = {
     countries: "Des pays",
     datasets: "Ensembles de données",
     directory: "Annuaire",
+    directoryEntries: "directory entries", // TODO: translate
     errorLoading: "FIXME",
     hideDirectory: "Masquer l’annuaire",
     in: "dans",
     loading: "Chargement des données",
     mapDisclaimer: "Cette carte contient des indications sur les zones où il y a des différends au sujet territorial. L'ACI n'approuve ni n'accepte les frontières représentées sur la carte.",
+    matchingResults: "matching results", // TODO: translate
```
Comment (wu-lee) [inline src/map-app/localisations.ts:154]: Maybe these translations be todone on this branch?

File: src/map-app/app/map-ui.ts [hunk 2]
```diff
@@ -131,10 +132,14 @@ export class MapUI {
...
     const isMulti = propDef.type === 'multi';
-    if (value === undefined)
+    if (value === undefined) {
       this.stateManager.clearPropFilter(propName);
-    else
+    } else {
       this.stateManager.propFilter(new PropEquality(propName, value, isMulti));
+    }
+
+    if (getViewportWidth() > 1080) // on smaller screens, wait until user clicks Apply Filters
```
Comment (wu-lee) [inline src/map-app/app/map-ui.ts:141]: I know this is a magic number from the stylesheets. But perhaps for self-documentation and to avoid them getting out of sync they should be a named constant?

Or, elsewhere I see the use of a utility function `getViewportWidth()`, which looks a bit like it's supposed to compute this value, perhaps that should be used?

File: src/map-app/app/map-ui.ts [hunk 3]
```diff
@@ -157,10 +162,13 @@ export class MapUI {
     // EventBus.Markers.needToShowLatestSelection.pub(lastContent.initiatives);
   }
 
-
   performSearch(text: string) {
     console.log("Search submitted: [" + text + "]");
     this.stateManager.textSearch(new TextSearch(text));
+
+    if (getViewportWidth() > 1080) {// on smaller screens, wait until user clicks Apply Filters
```
Comment (wu-lee) [inline src/map-app/app/map-ui.ts:164]: Ditto

File: src/map-app/app/view/sidebar.ts [hunk 4]
```diff
@@ -125,7 +120,7 @@ export class SidebarView extends BaseView {
     d3.select("#map-app-sidebar i").attr("class", "fa fa-angle-left");
 
     this.presenter.changeSidebar(); // Refresh the content of the sidebar
-    if (document.getElementById("dir-filter") && window.outerWidth >= 1080)
+    if (document.getElementById("dir-filter") && getViewportWidth() > 1080)
```
Comment (wu-lee) [inline src/map-app/app/view/sidebar.ts:124]: Ditto; although I see here perhaps `getViewportWidth()` computes something else.  Wondering, why was `outerWidth` not working adequately?

File: src/map-app/app/view/sidebar/directory.ts [hunk 5]
```diff
@@ -365,17 +181,15 @@ export class DirectorySidebarView extends BaseSidebarView {
...
       .append("i")
       .attr("class", "fa " + "fa-times");
     initiativeContentElement
       .append("div")
       .html(initiativeContent);
     initiativeSidebar.classed("sea-initiative-sidebar-open", true);
-    // if (document.getElementById("map-app-leaflet-map").clientWidth < 800)
-    if (window.outerWidth < 800)
+
+    if (getViewportWidth() <= 800)
```
Comment (wu-lee) [inline src/map-app/app/view/sidebar/directory.ts]: Another magic constant (not one of your creation of course, but same rationale as above applies, maybe)

File: src/map-app/app/view/sidebar/initiatives.ts [hunk 6]
```diff
@@ -56,25 +59,32 @@ export class InitiativesSidebarView extends BaseSidebarView {
     d3.select("#search-box").property("value", txt);
   }
 
-  createSearchBox(selection: d3DivSelection) {
+  submitCallback = (event: Event) => {
```
Comment (wu-lee) [inline src/map-app/app/view/sidebar/initiatives.ts:51]: Looks like this declaration lost its `const` qualifier, becoming a global?

File: src/map-app/app/map-ui.ts [hunk 7]
```diff
@@ -143,23 +143,18 @@ export class MapUI {
...
-    // Currently unimplemented.
+    const selectedInitiatives = this.mapPresenter?.getSelectedInitiatives() ?? [];
     
-    // EventBus.Markers.needToShowLatestSelection.pub(lastContent.initiatives);
+    if (selectedInitiatives.includes(initiative)) {
+      EventBus.Markers.needToShowLatestSelection.pub(selectedInitiatives.filter(i => i !== initiative));
+    } else {
+      EventBus.Markers.needToShowLatestSelection.pub([...selectedInitiatives, initiative]);
+    }
   }
```
Comment (wu-lee) [inline src/map-app/app/map-ui.ts:158]: Puzzled by this - seems to be re-implementing initiative selection, something which I had thought maybe used to be a thing, long ago but had been unused and probably broken for ages.

Doesn't seem to be related to counting and displaying visible initiatives?

File: src/map-app/app/presenter/map.ts [hunk 8]
```diff
@@ -8,7 +8,7 @@ import { MapUI } from '../map-ui';
 
 export class MapPresenter extends BasePresenter {
   readonly view: MapView;
-  private previouslySelected: Initiative[] = [];
+  private currentlySelected: Initiative[] = [];
```
Comment (wu-lee) [inline src/map-app/app/presenter/map.ts:12]: Puzzling name change. Can't see a change in the logic. Was the name wrong for the logic?

File: src/map-app/app/map-ui.ts
Comment (wu-lee) [inline src/map-app/app/map-ui.ts:1]: Puzzled why functionality from DirectorySidebarView should be moved into BaseSidebarView. Is this move deduplicating something?

File: src/map-app/localisations.ts (hunk 1 above)
Comment (rogup) [reply src/map-app/localisations.ts:154]: @ColmDC What's the process for doing these translations? I think I could manage French and Spanish using Google Translate, Linguee, and my knowledge from GCSE, but I wouldn't feel confident in my Korean translations.

File: src/map-app/app/map-ui.ts (hunk 2 above)
Comment (rogup) [reply src/map-app/app/map-ui.ts:141]: Yep I'll add a constant, in both JS and CSS

File: src/map-app/app/view/sidebar.ts (hunk 4 above)
Comment (rogup) [reply src/map-app/app/view/sidebar.ts:124]: Outerwidth includes other browser layout, like the scrollbar and dev tools. Inner width is more accurate but not exactly the same as @media in css, so this is just to make it consistent.

File: src/map-app/app/view/sidebar/directory.ts [hunk 9]
```diff
@@ -68,11 +68,11 @@ export class DirectorySidebarView extends BaseSidebarView {
       d3.selectAll("li.sea-directory-field").attr("hidden", null);
     else {
       const a = d3.selectAll("li.sea-directory-field");
-      a.attr("hidden", null); // Hide everything
+      a.attr("hidden", true); // Hide everything
       a.filter(function () {
         // Unhide elements which have matching text
         return d3.select(this).text().toLowerCase().includes(input);
-      }).attr("hidden", true);
+      }).attr("hidden", null);
```
[the remark this answers is not in the document]
Comment (rogup) [reply src/map-app/app/view/sidebar/directory.ts:83]: I'm not even sure if this feature is meant to exist, because it's hidden in the UI and hard to find it with the cursor (just below the Directory header).

But I made this change because it was broken and it was an easy fix. The old behaviour was in reverse:
`a.attr("hidden", null);` unhid everything and then the ones that matched were actually hidden by
```
a.filter(function () {
        // Unhide elements which have matching text
        return d3.select(this).text().toLowerCase().includes(input);
      }).attr("hidden", true);
```

Swapping the nulls and trues reversed the behaviour to match what the code comments say.

File: src/map-app/app/view/sidebar/initiatives.ts (hunk 6 above)
Comment (rogup) [reply src/map-app/app/view/sidebar/initiatives.ts:51]: I pulled it out of the scope of the `createSearchBox` method because I thought I needed to use it elsewhere in the class, but I didn't actually need to. I'll add it back.

File: src/map-app/app/map-ui.ts (hunk 7 above)
Comment (rogup) [reply src/map-app/app/map-ui.ts:158]: This method wasn't implemented, but I needed it in order to highlight the selected initiative in the sidebar.

The previous code for populating the initiative sidebar tried to achieve this by setting the class `.sea-initiative-active` but this didn't work because the class didn't persist when the sidebar UI was refreshed.

File: src/map-app/app/map-ui.ts
Comment (rogup) [reply src/map-app/app/map-ui.ts:1]: Because as part of this feature's requirements, we need to reuse the same results pane UI for InitiativesSidebarView, so it needed to be moved up a level from DIrectorySidebarView.

File: src/map-app/localisations.ts (hunk 1 above)
Comment (ColmDC) [reply src/map-app/localisations.ts:154]: Can you specify the mising terms?

File: src/map-app/app/view/sidebar/directory.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (wu-lee) [reply src/map-app/app/view/sidebar/directory.ts:83]: I think it's a vestige.  Maybe to be removed, but I've been cautious about just removing stuff without understanding. (My understanding is still often incomplete.)

Re-reading, ignore points 3 - 5, I think it does invert correctly.

File: src/map-app/app/map-ui.ts (hunk 7 above)
Comment (wu-lee) [reply src/map-app/app/map-ui.ts:158]: I'm uncertain, but I seem to remember that selection stuff is to do with selecting markers on the map, not selecting sidebar items. 

I think here you're just using them to zoom into the marker? Via `EventBus.Markers.needToShowLatestSelection`?

Not a big deal, just checking one sort of functionality isn't being confused for another and revived to serve a different purpose.

**Comment by wu-lee** [comment]
> @wu-lee I can have a go at rebasing the code to make it easier for you. I think this would be be best practise, but trying to do so felt daunting because the code was quite messy so it would take a while for me to untangle the commits, and I don't know if this was worth the effort.
> 
> But I think it would be good for me to get more confident in rebasing anyway, and I'll need to do it for the PR, unless I do one massive merge. Could we have a call where you show me how you do rebasing?

Sure - we can compare methods. I'm a bit sketchy about how to do it with VSCode/GitLens, mostly I do it with Emacs/Magit/CLI, but would be interested in working out how to.

File: src/map-app/app/map-ui.ts (hunk 7 above)
Comment (rogup) [reply src/map-app/app/map-ui.ts:158]: I think because I'm coming from LX, I'm seeing things in a Redux "app state" kind of way.

We sort of keep a state of the selected markers on the map, within the `currentlySelected` array below. When we click on an initiative in the sidebar, or on a marker directly, it adds the initiative to this array. So I've written this function to expose this state, so that the currently selected initiative is also highlighted in the results pane. It should work however the marker is selected.

**Tangential to this point:** I think this is actually the most frustrating bit of the UI. The app's state is fragmented and stored all over the code, within different classes. And then if another class wants to access or change the state, it's a nightmare to thread the information through a chain of functions, especially if they're in different parts of the UI.
It's a limitation of OOP that I experienced as a Java Android developer, since it often leads to anti-patterns, unless the code is written in a very strict and methodical way with a clear structure (which this isn't).
If we were to re-write some of the UI code, I think this might actually be the most beneficial change- to move app state to a central store such as Redux, which any part of the UI can hook into. Since we implemented Redux more widely in LX, it has made the code a lot cleaner. I think this is more pressing than changing to something like React. The html structure is already fairly understandable and d3 doesn't seem terrible.

File: src/map-app/localisations.ts (hunk 1 above)
Comment (rogup) [reply src/map-app/localisations.ts:154]: EN: Apply Filters
FR: Appliquer les filtres
ES: Aplicar filtros
KO: 필터 적용

EN: <number> directory entries
FR: <number> entrées d'annuaire
ES: <number> entradas de directorio
KO: <number> 디렉토리 항목

EN: <number> matching results
FR: <number> résultats correspondants
ES: <number> resultados coincidentes
KO: <number> 일치하는 결과

File: src/map-app/app/view/sidebar/directory.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (rogup) [reply src/map-app/app/view/sidebar/directory.ts:83]: Ok I'll keep it for now then

File: src/map-app/app/map-ui.ts (hunk 2 above)
Comment (rogup) [reply src/map-app/app/map-ui.ts:141]: So... I've done it for JS but it's not actually possible to set named constants for css @media queries yet. https://stackoverflow.com/questions/40722882/css-native-variables-not-working-in-media-queries

There are some hacky workarounds, or we could use some npm libraries to enhance the standard css functionality, but I think that's out of scope for this issue. We only use the constants in 3 places in our CSS stylesheet, so I think it's something we can live with for now.

File: src/map-app/localisations.ts (hunk 1 above)
Comment (wu-lee) [reply src/map-app/localisations.ts:154]: Wondering about the "N"s - partly if they translate literally to other languages, partly if it's a widely understood layman's term to use in English, and partly, now I look in the diff, if it's just a mistake which isn't in the original phrase? 

I suppose there's a question about where the number comes in the other language phrases, in particular Korean. I know nothing useful about Korean...

Comment (rogup) [reply src/map-app/localisations.ts:154]: The N isn't in the translation, it's just "directory entries". I included the N as context that the string will come after a number.
According to Google translate, Korean doesn't always put the number before the string, but I'm thinking we just ignore this and see if anyone complains? To do it properly, we would have to improve the way we do localisation by including context comments and a way to do string interpolation.

**Comment by rogup** [comment]
@wu-lee I've rebased the commits as best as I could, but there's still probably a bit of messiness unless I spent many more hours.

So I think all that's left, unless you have any more review comments, is to wait on confirmation about the translations. Then I'll send it over to Colm for QA.

File: src/map-app/app/map-ui.ts (hunk 7 above)
Comment (wu-lee) [reply src/map-app/app/map-ui.ts:158]: I think something like this would be a first step to anything similar to introducing react anyway.

However, suspect it still won't be easy! There is a lot of stuff, and untangling all the state might not be simple. Therefore I suspect it would need to be further broken down into steps which can be attempted independently...

I've tried to get rid of the EventBus stuff as much as I could and at least make it typesafe (previously it was easy to send malformed messages to misnamed targets, and what the allowed targets was totally opaque and undocumented).  I've not entirely succeeded but it's better than it was.  What probably needs to be done is to continue with that so that we have a kind if "central source of truth" which behave more like you get in Redux. But there are other bits of state and coupled logic all over the place, and I don't know how deep those depths go...

Which ultimately means we need to get budget-holder clearance to put our snorkels on and try investigating - but leave a route to bail out if it starts getting too tangled or taking too long.

On the other hand - we might want to look at the bigger picture an see if that goes somewhere we need to be, or if we can instead dredge out the useful bits and retrofit them into a new framework - perhaps if we want to be able to make very big maps?   Therefore needs the kind of high-level outcomes planning we scheduled (for today but had to postpone.)

**Comment by wu-lee** [comment]
I think it's probably just as well to pick option 1, yes.

**Comment by ColmDC** [comment]
Does this need to be in the In Review column? 

is there anything blocking my seeing this in action?

**Comment by rogup** [comment]
@ColmDC Please could you confirm these translations with ICA https://github.com/DigitalCommons/mykomap/pull/246#discussion_r1583450832

Then I'll write them in and send it to you for QA

**Comment by ColmDC** [comment]
Just add google translations for them and send to QA. I'm keen to close this one.

**Comment by wu-lee** [comment]
The above merge looks like a fast-forward merge?  i.e. it looks like the branch has been flattened into `dev`.. compare with `230-initiative-filtering-ui` on github currently which has identical commits.
